javaparser-core/src/main/java/com/github/javaparser/printer/configuration/PrinterConfiguration.java

```diff
@@ -0,0 +1,216 @@
+ * b) the terms of the Apache License
+ *
+ * You should have received a copy of both licenses in LICENCE.LGPL and
+ * LICENCE.APACHE. Please refer to those files for details.
+ *
+ * JavaParser is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ */
+
+package com.github.javaparser.printer.configuration;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+import com.github.javaparser.printer.Printable;
+import com.github.javaparser.printer.configuration.Indentation.IndentType;
+import com.github.javaparser.utils.Utils;
+
+/**
+ * Configuration options for the {@link Printable}.
+ */
+public class PrinterConfiguration implements ConfigurablePrinter {
+    
+    public enum ConfigOption {
+        ORDER_IMPORTS(Boolean.class), 
+        PRINT_COMMENTS(Boolean.class), 
+        PRINT_JAVADOC(Boolean.class), 
+        SPACE_AROUND_OPERATORS(Boolean.class), 
+        COLUMN_ALIGN_PARAMETERS(Boolean.class), 
+        COLUMN_ALIGN_FIRST_METHOD_CHAIN(Boolean.class),
+        /**
+         * Indent the case when it is true, don't if false
+         * switch(x) {            switch(x) {
+         *    case 1:             case 1:
+         *        return y;           return y;
+         *    case 2:             case 2:
+         *        return z;           return x;
+         *}                       }
+         */
+        INDENT_CASE_IN_SWITCH(Boolean.class),
+        /**
+         * By default enum constants get aligned like this:
+         * <pre>
+         *     enum X {
+         *        A, B, C, D
+         *     }
+         * </pre>
+         * until the amount of constants passes this value (5 by default).
+         * Then they get aligned like this:
+         * <pre>
+         *     enum X {
+         *        A,
+         *        B,
+         *        C,
+         *        D,
+         *        E,
+         *        F,
+         *        G
+         *     }
+         * </pre>
+         * Set it to a large number to always align horizontally.
+         * Set it to 1 or less to always align vertically.
+         */
+        DEFAULT_MAX_ENUM_CONSTANTS_TO_ALIGN_HORIZONTALLY(Integer.valueOf(5)),
+        END_OF_LINE_CHARACTER(Utils.SYSTEM_EOL);
+        
+        Object value;
+        
+        Class type;
+        
+        // Option without value
+        <T> ConfigOption(Class clazz) {
+            this.type = clazz;
+        }
+        
+        // Option with initial value
+        ConfigOption(Object value) {
+            value(value);
+        }
+        
+       /* 
+        * Set a value to an option
+        */
+        public ConfigOption value(Object value) {
+            Utils.assertNotNull(value);
+            this.value = value;
+            this.type = value.getClass();
```

Comment:
I'm rusty with generics -- If the enum is initialised as being e.g. `Boolean.class` and it turns out that `value` is an `Integer`, will this overwrite the type?

Reply:
I haven't tested but it should throw a ClassCastException exception

Reply:
ConfigurablePrinter config = new PrinterConfiguration();
config.addOption(ConfigOption.ORDER_IMPORTS.value("1")); // there is no verification here
Boolean b1 = config.get(ConfigOption.ORDER_IMPORTS).get().asBoolean(); // java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Boolean
Boolean b2 = config.get(ConfigOption.ORDER_IMPORTS).get().asValue(); // java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Boolean
Perhaps it could be improved!

Reply:
Now I implements a fail fast assignment
For example if ConfigOption.PRINT_COMMENTS is configured with a Boolean type
config.addOption(ConfigOption.PRINT_COMMENTS.value("1")); // throws an IllegalArgumentException

Reply:
PrinterConfiguration config = new DefaultPrinterConfiguration();
config.addOption(new ConfigurationOption(ConfigOption.PRINT_COMMENTS, "1")); // throws an IllegalArgumentException